extensions/amp-analytics/0.1/events.js

```diff
@@ -436,34 +695,101 @@ export class TimerEventTracker extends EventTracker {
-        'Bad maxTimerLength specification');
-    const callImmediate = 'immediate' in timerSpec ?
-        Boolean(timerSpec['immediate']) : true;
-
-    const win = this.root.ampdoc.win;
-    const intervalId = win.setInterval(() => {
-      listener(this.createEvent_(eventType));
-    }, interval * 1000);
-    this.trackers_.push(intervalId);
-    win.setTimeout(() => {
-      this.removeTracker_(intervalId);
-    }, maxTimerLength * 1000);
-    if (callImmediate) {
-      listener(this.createEvent_(eventType));
+    const timerStart = 'startSpec' in timerSpec ? timerSpec['startSpec'] : null;
+    user().assert(!timerStart || typeof timerStart == 'object',
+        'Bad timer start specification');
+    const timerStop = 'stopSpec' in timerSpec ? timerSpec['stopSpec'] : null;
+    user().assert((!timerStart && !timerStop) || typeof timerStop == 'object',
+        'Bad timer stop specification');
+
+    const timerId = this.generateTimerId_();
+    let startBuilder;
+    let stopBuilder;
+    if (!!timerStart) {
+      const startTracker = this.getTracker(timerStart);
+      user().assert(startTracker, 'Cannot track timer start');
+      startBuilder = startTracker.add.bind(startTracker, context,
+          timerStart['on'], timerStart,
+          this.handleTimerToggle_.bind(this, timerId, eventType, listener));
     }
+    if (!!timerStop) {
+      const stopTracker = this.getTracker(timerStop);
+      user().assert(stopTracker, 'Cannot tracker timer stop');
+      stopBuilder = stopTracker.add.bind(stopTracker, context,
+          timerStop['on'], timerStop,
+          this.handleTimerToggle_.bind(this, timerId, eventType, listener));
+    }
+
+    const timerHandler = new TimerEventHandler(
+        timerSpec, startBuilder, stopBuilder);
+    this.trackers_[timerId] = timerHandler;
+
+    timerHandler.init(
+        this.startTimer_.bind(this, timerId, eventType, listener));
     return () => {
-      this.removeTracker_(intervalId);
+      this.removeTracker_(timerId);
     };
   }
 
+  /**
+   * @return {number}
+   * @private
+   */
+  generateTimerId_() {
+    return ++this.timerIdSequence_;
+  }
+
+  /**
+   * @param {!JsonObject} config
+   * @return {?EventTracker}
+   */
+  getTracker(config) {
+    const eventType = user().assertString(config['on']);
+    const trackerKey = getTrackerKeyName(eventType);
+
+    return this.root.getTrackerForWhitelist(
+        trackerKey, getTrackerTypesForParentType('timer'));
+  }
+
+  /**
+   * Toggles which listeners are active depending on timer state, so no race
+   * conditions can occur in the case where the timer starts and stops on the
+   * same event type from the same target.
+   * @param {number} timerId
+   * @param {string} eventType
+   * @param {function(!AnalyticsEvent)} listener
+   * @private
+   */
+  handleTimerToggle_(timerId, eventType, listener) {
+    const timerHandler = this.trackers_[timerId];
+    if (timerHandler.isRunning()) {
```

Comment:
I don't think this will work properly in special case that stopEvent happens first. I think we need an extra boolean param indicating start/stop instead. 
And we won't need a `#isRunning` method then.

Reply:
nvm, this should never happen.

Reply:
Yep, that's why we need the toggle, we only listen for one type of event at a time to eliminate race conditions like that. :)